mshv-ioctls/src/ioctls/vm.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright © 2020, Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+use crate::ioctls::vcpu::{new_vcpu, VcpuFd};
+use crate::ioctls::Result;
+use crate::mshv_ioctls::*;
+use mshv_bindings::*;
+use std::fs::File;
+
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use vmm_sys_util::errno;
+use vmm_sys_util::ioctl::{ioctl_with_mut_ref, ioctl_with_ref};
+
+/// Wrapper over Mshv VM ioctls.
+pub struct VmFd {
+    vm: File,
+}
+
+impl AsRawFd for VmFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.vm.as_raw_fd()
+    }
+}
+
+impl VmFd {
+    ///
+    /// Install intercept to enable some VM exits like MSR, CPUId etc
+    ///
+    pub fn install_intercept(&self, install_intercept_args: mshv_install_intercept) -> Result<()> {
+        #[allow(clippy::cast_lossless)]
+        let ret =
+            unsafe { ioctl_with_ref(self, MSHV_INSTALL_INTERCEPT(), &install_intercept_args) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+    ///
+    /// Creates/modifies a guest physical memory.
+    ///
+    pub fn map_user_memory(&self, user_memory_region: mshv_user_mem_region) -> Result<()> {
+        #[allow(clippy::cast_lossless)]
+        let ret = unsafe { ioctl_with_ref(self, MSHV_MAP_GUEST_MEMORY(), &user_memory_region) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+    ///
+    /// Unmap a guest physical memory.
+    ///
+    pub fn umap_user_memory(&self, user_memory_region: mshv_user_mem_region) -> Result<()> {
+        #[allow(clippy::cast_lossless)]
+        let ret = unsafe { ioctl_with_ref(self, MSHV_UNMAP_GUEST_MEMORY(), &user_memory_region) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+    ///
+    /// Creates a new MSHV vCPU file descriptor
+    ///
+    pub fn create_vcpu(&self, id: u8) -> Result<VcpuFd> {
+        let vp_arg = mshv_create_vp {
+            vp_index: id as __u32,
+        };
+        // Safe because we know that vm is a VM fd and we verify the return result.
+        #[allow(clippy::cast_lossless)]
+        let vcpu_fd = unsafe { ioctl_with_ref(&self.vm, MSHV_CREATE_VP(), &vp_arg) };
+        if vcpu_fd < 0 {
+            println!("vcpu fd: {}", vcpu_fd);
+            return Err(errno::Error::last());
+        }
+
+        // Wrap the vCPU now in case the following ? returns early. This is safe because we verified
+        // the value of the fd and we own the fd.
+        let vcpu = unsafe { File::from_raw_fd(vcpu_fd) };
+
+        Ok(new_vcpu(vcpu))
+    }
+    ///
+    /// Creates an interrupt controller.
+    ///
```

Comment:
Is this really creating an interrupt controller or rather allocating/requesting an interrupt vector?

There are quite a few arguments to this function, some of them rather obscure. It's usually a good idea to limit the number of public function arguments in order to make the call easier by simply passing a wrapper structure instead.
So I'd suggest to create such structure, make it public, document its fields and have `request_virtual_interrupt` take `&self` and `&interrupt_config` as arguments.

Reply:
The comment here is wrong. This function does not create an interrupt controller. It is used to inject an interrupt into the guest.

Reply:
Changed to comment. I will create an issue to make the parameter changes in following PR.

Reply:
https://github.com/cloud-hypervisor/mshv/issues/3